### DISCUSSION TITLE
Volume + Population

### DISCUSSION THREAD

**Discussion by AdrianKriger** [liberostelios/urban-morphology-3d#14] [Q&A]
As always: your work is pretty awesome. Thank you.

- Your input needs to parse through val3dity successfully. Is this correct?
No non-manifold (missing vertices) / polygon wrong direction errors? 

I have failed to create a topologically correct model. Since I want to calculate metrics for a ‘_primitive_’ LoD1 dataset are their other methods to determine a volume without ‘_bucky_’ libraries? 

Would `scipy.spatial.ConvexHull(points).volume` [[volume.py line 80](https://github.com/tudelft3d/3d-building-metrics/blob/main/volume.py#L80)] do the business? 
What about shared walls / nearest neighbour?

- Is there a _really easy_ method to estimate population, for a given neighbourhood, based on building=use and building=level? How? Can you point me in the right direction please?

**Comment by liberostelios** [comment]
Hey @AdrianKriger,

Thank you so much 🙌

> Your input needs to parse through val3dity successfully. Is this correct?

No, that's not necessary. This is, in particular, because certain statistics were build to be reliable despite the input's validity. For instance, we compute the AABB,  OBB and convex hull volume already, which are going to be correct despite the topological correctness of the model.

As per [here](https://github.com/tudelft3d/3d-building-metrics#tutorial-please), you can run val3dity against your dataset before. and then pass that report as an argument if you want to inherit this information to the statistics. We implemented that in particular as we wanted to run stats on all buildings, but then be able to know for which we can trust all metrics and for which it makes sense to only look at the subset of robust metrics (like the volumes mentioned before).

> Is there a really easy method to estimate population, for a given neighbourhood, based on building=use and building=level? How? 

Can you point me in the right direction please?
I think @Athelena and @fbiljecki take this one. I am only wondering if you are referring to OSM properties here.

**Reply by AdrianKriger** [reply]
Yes @liberostelios . The OSM `key:value` pair. 

I want to give a user [[-> here](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/interactiveOnly.ipynb)] the ability to estimate population for a given community.

To prevent calling OSM (overpass) the notebook might change; and harvest the data from the previously created 3D.cityjson. How challenging is it to transform a 3D.cityjson to a `gdf`? I see you can go to a `df` but I need the `geometry` (footprint Polygon) please. Do you recommend going this route? 

Either way: `volume`, `building=levels` + `=use` would be the data available.

**Reply by liberostelios** [reply]
I believe what you are looking for is [this](https://github.com/tudelft3d/3d-building-metrics/blob/ac8712d3c876b38ff2a5ca5deb9ebc5118c90d8b/cityjson.py#L28) function? Feel free to reuse that. 😉

Fwiw, I checked your referred notebook and it's pretty cool stuff! What I would have probably done (not that it would necessarily be for better) is to convert OSM data to a `gdf` and do my analysis with them. But it's all about what works for you.

**Reply by AdrianKriger** [reply]
You are very kind @liberostelios.
 
Let me understand what you are saying:
Harvest OSM. Transform `building=levels` to building height -> go to `gdf` then `use` + `volume` and estimate population. 
Is that what you mean?
It does make sense. 

Perhaps it would be best to keep that notebook, with improvements to make it less clunky, for those who want a visualization and some basic spatial analysis. 
And do another that harvests from the 3D.CityJSON. 
I want to keep these separate and call OSM the least number of times [_be less of a burden_]; for each desired product. 
So its; call once = visualization + estimate population. 
Or call once = LoD1 3D CItyModel + 2.75D surface mesh + visualization + estimate population. 

I’ll get to that. First; I’m trying to get the CityJSON topologically correct and  its surface mesh with the least amount of errors possible [-> here]( https://github.com/cityjson/misc-example-code/issues/1)

**Comment by AdrianKriger** [comment]
Is this what you had in mind @liberostelios [-> here](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/interactiveOnly.ipynb)?

While nowhere near the quality and scale of @Athelena and @fbiljecki work; I would really appreciate their direction and advice.  
*---should I test it more widely? What have I missed / overlooked? What should I consider?*

**Reply by AdrianKriger** [reply]
@liberostelios I just realized; in order to ensure the quality of the result the input building footprints cannot cross. (i.e.: need to be as correct as possible). 
I know these footprints are topologically correct but many areas won't be. To go further [spatial analysis] a topology test similar to the one in [osm_Lod1_3DCityModel-walkthrough](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/extra/osm_Lod1_3DCityModel-walkthrough.ipynb) is necessary.

|  |  |
| --- | --- |
|![cput-error](https://github.com/tudelft3d/3d-building-metrics/assets/59996720/0af80c67-7712-48ae-895a-1d1bf637e11e)  | Challenges will highlight in _'RED'_ and need investigation |
|![cput](https://github.com/tudelft3d/3d-building-metrics/assets/59996720/60507570-367f-4eaa-9e64-3093dfe48fbe) | or not  |

Am I correct? 
mmm... If this is correct the: call once -> 3D.cityjson -> interactive viz -> spatial data science would be the preferred process. 
Can you confirm please? 
If yes; can you please direct me towards the most effective / efficient way to harvest the building footprints, from a 3D.CityJSON as a `geometry` column in a `gdf`.

**Comment by liberostelios** [comment]
Sorry @AdrianKriger, I noticed I missed many of your previous posts.

I have no opinion on how to estimate the population explicitly. My suggestions are purely around the geometric manipulation of the data.

If you have a 3D.cityjson and you load it somewhere, you can simply iterate through all the CityObjects and create a dictionary and then use a [`from_dict`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.from_dict.html) function to create the dataframe. Something like this:

```Python
import pandas as pd
from geopandas import GeoDataFrame

# Assuming cm is your city model 
data = {}

for obj_id in cm["CityObjects"]:
    data[obj_id] = {
        **data[obj_id]["attributes"],
        "geometry": to_shapely(data[obj_id]["geometry"][0], cm["vertices"], ground_only="True")
    }

gdf = GeoDataFrame(pd.DataFrame.from_dict(data, orient="rows"))
```

The code is a little rough and might need some fixing here and there (I didn't test it) but should work.

**Reply by AdrianKriger** [reply]
Thank you.

I do not have semantic surfaces (only LoD1) and created a function  [-> [here](https://github.com/AdrianKriger/osm_LoD1_3DCityModel/blob/main/village_campus/CityJSONspatialDataScience.ipynb)] but its not harvesting the correct data. When a building has interiors (holes / courtyards) it grabs those geometries and not the outer shell.

Can I add semantic surfaces to LoD1? It would make harvesting the geometry easier.

**Comment by AdrianKriger** [comment]
I carried the footprint from OpenStreetMap through the CityJSON, as an attribute, and then harvested that.
